test/unittest/datasets/unittest_func_data_producers.cpp

```
@@ -0,0 +1,102 @@
+  /** test user data is given correctly */
+  int *ud = reinterpret_cast<int *>(user_data);
+  *ud += 1;
+
+  /** first input/label is all zero, second input/label is all one */
+  auto first_input = nntrainer::Tensor::Map(
+    *outVec, input_shapes[0].getDataLen(), input_shapes[0]);
+  first_input.setValue(0);
+
+  auto second_input = nntrainer::Tensor::Map(
+    *(outVec + 1), input_shapes[1].getDataLen(), input_shapes[1]);
+  second_input.setValue(1);
+
+  auto first_label = nntrainer::Tensor::Map(
+    *outLabel, label_shapes[0].getDataLen(), label_shapes[0]);
+  first_label.setValue(0);
+
+  auto second_label = nntrainer::Tensor::Map(
+    *(outLabel + 1), label_shapes[1].getDataLen(), label_shapes[1]);
+  second_label.setValue(1);
+  *last = false;
+
+  return 0;
+};
+
+int getBatch_error(float **outVec, float **outLabel, bool *last,
+                   void *user_data) {
+  return -1;
+}
+
+bool validate(const std::vector<nntrainer::Tensor> &inputs,
+              const std::vector<nntrainer::Tensor> &labels) {
+  if (user_data == 0 || inputs.size() != 2 || labels.size() != 2) {
+    return false;
+  }
+
+  nntrainer::Tensor expected_first_input(input_shapes[0]);
+  expected_first_input.setValue(0);
+  nntrainer::Tensor expected_second_input(input_shapes[1]);
+  expected_second_input.setValue(1);
+
+  nntrainer::Tensor expected_first_label(label_shapes[0]);
+  expected_first_label.setValue(0);
+  nntrainer::Tensor expected_second_label(label_shapes[1]);
+  expected_second_label.setValue(1);
+
+  return inputs[0] == expected_first_input &&
+         inputs[1] == expected_second_input &&
+         labels[0] == expected_first_label &&
+         labels[1] == expected_second_label;
+};
+
+} // namespace
+
+std::unique_ptr<nntrainer::DataProducer>
+createConstantBatchProducer(const std::vector<std::string> &properties = {}) {
+  std::unique_ptr<nntrainer::DataProducer> ptr =
+    std::make_unique<nntrainer::FuncDataProducer>(getBatch, &user_data);
```

can you make another unit test when you pass nullptr to this getBatch, and let it throw/fail and see our system handles it.

---

Added